.github/workflows/ci.yml

```diff
@@ -31,19 +31,31 @@ jobs:
     env:
       MAX_COUNT: 6
     steps:
-      - run: echo "FETCH_DEPTH=$((${{ github.event.pull_request.commits }} + 2))" >> "$GITHUB_ENV"
+      - name: Determine fetch depth
+        run: echo "FETCH_DEPTH=$((${{ github.event.pull_request.commits }} + 2))" >> "$GITHUB_ENV"
       - uses: actions/checkout@v4
         with:
           ref: ${{ github.event.pull_request.head.sha }}
           fetch-depth: ${{ env.FETCH_DEPTH }}
-      - run: |
+      - name: Determine commit range
+        run: |
+          # Checkout HEAD~ and find the test base commit
+          # Checkout HEAD~ because it would be wasteful to rerun tests on the PR
+          # head commit that are already run by other jobs.
           git checkout HEAD~
-          echo "COMMIT_AFTER_LAST_MERGE=$(git log $(git log --merges -1 --format=%H)..HEAD --format=%H --max-count=${{ env.MAX_COUNT }} | tail -1)" >> "$GITHUB_ENV"
+          # Figure out test base commit by listing up to MAX_COUNT ancestors of
+          # HEAD in order by timestamp, excluding ancestors of the most recent
+          # merge commit, and taking the first one. If branch contains up to
+          # MAX_COUNT ancestor commits after the last merge commit, all of those
+          # commits will be tested. If it contains more, only the most recent
+          # MAX_COUNT commits will be tested.
+          echo "TEST_BASE=$(git rev-list -n$((${{ env.MAX_COUNT }} + 1)) --reverse HEAD ^$(git rev-list -n1 --merges HEAD)^@ | head -1)" >> "$GITHUB_ENV"
```

Comment:
re: https://github.com/bitcoin/bitcoin/pull/28573#discussion_r1346049641

> (Can be written shorter)

This is a little shorter, but I prefer head instead of tail in cases like this because the rev-list command can detect when the pipe is closed and avoid generating unnecessary output.